File: code/js/src/Domain/invitations/ImInvitation.ts
import { ImInvitationOutputModel } from '../../Dto/output/invitations/ImInvitationOutputModel';

/**
 * Represents an invitation to create an instant messaging account.
 *
 * @param token The one-time use token.
 * @param expiresAt The date and time when the token expires.
 */
export interface ImInvitation {
    token: string;
    expiresAt: Date;
}

export function imInvitationFromDto(
    dto: ImInvitationOutputModel,
): ImInvitation {
    return {
        token: dto.token,
        expiresAt: new Date(dto.expiresAt),
    };
}

File: code/server/jvm/im-domain/src/main/kotlin/im/domain/invitations/ImInvitation.kt
package im.domain.invitations

import java.time.LocalDateTime
import java.util.UUID

/**
 * Represents an invitation to a channel.
 * @property token The unique identifier of the invitation.
 * @property status The status of the invitation.
 * @property expiresAt The date and time when the invitation expires.
 * @property expired Indicates if the invitation has expired.
 */
data class ImInvitation(
    val token: UUID = UUID.randomUUID(),
    val status: ImInvitationStatus = ImInvitationStatus.PENDING,
    val expiresAt: LocalDateTime,
) {
    val expired: Boolean
        get() = expiresAt.isBefore(LocalDateTime.now())

    fun use(): ImInvitation = copy(status = ImInvitationStatus.USED)
}
